xds/client.go

```diff
@@ -0,0 +1,277 @@
+}
+
+func (c *Client) sendRequest(resourceNames []string, version, nonce, errMsg string) error {
+	req := &v3discoverypb.DiscoveryRequest{
+		Node:          c.nodeProto,
+		TypeUrl:       resource.EndpointType,
+		ResourceNames: resourceNames,
+		VersionInfo:   version,
+		ResponseNonce: nonce,
+	}
+	if errMsg != "" {
+		req.ErrorDetail = &statuspb.Status{
+			Code: int32(codes.InvalidArgument), Message: errMsg,
+		}
+	}
+
+	if err := c.stream.Send(req); err != nil {
+		return fmt.Errorf("xds: stream.Send(%+v) failed: %v", req, err)
+	}
+
+	return nil
+}
+
+func (c *Client) run(ctx context.Context) {
+	retries := 0
+	streamRunning := true
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(c.done)
+
+			return
+		default:
+		}
+
+		if retries != 0 {
+			timer := time.NewTimer(c.strategy.Backoff(retries))
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				if !timer.Stop() {
+					<-timer.C
+				}
+
+				return
+			}
+		}
+
+		retries++
+
+		//restart if xds stream is not already running
+		if !streamRunning {
+			//Do not use c.Start here as it calls run() again
+			if err := c.restart(ctx); err != nil {
+				//ToDo: Consider exposing metrics to track this
+				c.logger.Error(err, "xds: Failure to recover from broken stream:", "retries", retries)
+
+				continue
+			}
+
+			retries = 0
+		}
+
+		if c.recv() {
+			retries = 0
+		}
+
+		streamRunning = false
+	}
+}
+
+func (c *Client) parseResponse(r proto.Message) ([]*anypb.Any, string, string, error) {
+	resp, ok := r.(*v3discoverypb.DiscoveryResponse)
+	if !ok {
+		return nil, "", "", fmt.Errorf("xds: unsupported message type: %T", resp)
+	}
+
+	// Note that the xDS transport protocol is versioned independently of
+	// the resource types, and it is supported to transfer older versions
+	// of resource types using new versions of the transport protocol, or
+	// vice-versa. Hence we need to handle v3 type_urls as well here.
+	var err error
+
+	url := resp.GetTypeUrl()
+
+	if url != resource.EndpointType {
+		return nil, "", "", fmt.Errorf("xds: resource type (%v) is not EndpointResource in server response",
+			resp.GetTypeUrl())
+	}
+
+	return resp.GetResources(), resp.GetVersionInfo(), resp.GetNonce(), err
+}
+
+func (c *Client) nack(err error) {
+	if e := c.sendRequest([]string{c.xdsTarget}, c.vsn, c.nonce, err.Error()); e != nil {
+		c.logger.Error(e, "xds: Nack: SendRequest error", "version", c.vsn, "nonce", c.nonce)
+	}
+}
+
+func (c *Client) ack() {
+	if err := c.sendRequest([]string{c.xdsTarget}, c.vsn, c.nonce, ""); err != nil {
+		c.logger.Error(err, "xds: Ack: SendRequest error", "version", c.vsn, "nonce", c.nonce)
+	}
+}
+
+func (c *Client) recv() bool {
+	success := false
+
+	for {
+		c.logger.Debug("xds: Restarting recv loop ")
+
+		resp, err := c.stream.Recv()
+
+		if err != nil {
+			c.logger.Error(err, "xds: error while recv()")
+
+			return success
+		}
+
+		resources, vsn, nonce, err := c.parseResponse(resp)
+
+		if err != nil {
+			c.nack(err)
+
+			success = true
```

Comment:
should we set this as success here ? as we are sending nack back

Reply:
success is just being used to determine if we were able to successfully receive any data from the management server. It helps in determining if we need to reconnect/restart the stream.

Reply:
We should rename this then to convey the intent clearly.

Reply:
Renamed it to isReceived